File: solution/077Combinations/077Combinations.go
package main

import "fmt"

func backtracking(n int, target int, k int, results [][]int, a []bool) [][]int {
	if k == n-1 {
		c := 0
		for i := 0; i < n; i++ {
			if a[i] {
				c++
			}
		}
		if c == target {
			result := make([]int, target)
			c = 0
			for i := 0; i < n; i++ {
				if a[i] {
					result[c] = i + 1
					c++
				}
			}
			results = append(results, result)
		}
	} else {
		k++
		cands := []bool{false, true}
		for i := 0; i < len(cands); i++ {
			a[k] = cands[i]
			results = backtracking(n, target, k, results, a)
		}
	}
	return results
}
func combine(n int, k int) [][]int {
	var results [][]int
	a := make([]bool, n)
	results = backtracking(n, k, -1, results, a)
	return results
}

func main() {
	fmt.Printf("%v\n", combine(4, 2))
}

File: src/077Combinations/077Combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> results;
        vector<bool> a;
        backtracking(a, 0, n, k, results);
        return results;
    }
    
    void backtracking(vector<bool>&a, int i, int n, int k, vector<vector<int>>& results){
        if(is_solution(a, i, n, k)){
            process_solution(a, i, n, k, results);
        }else if(i<n){
            auto cands = construct_candidates(a, i, n, k);
            ++i;
            for(const auto& cand : cands) {
                a.push_back(cand);
                backtracking(a, i, n, k, results);
                a.pop_back();
            }
        }
    }
    
    bool is_solution(vector<bool>&a, int i, int n, int k) {
        return i==n && k == std::count(a.begin(), a.end(), true);
    }
    
    void process_solution(vector<bool>&a, int i, int n, int k, vector<vector<int>>& results) {
        vector<int> result;
        for(int j=0; j<n; ++j){
            if(a[j]) result.push_back(j+1);
        }
        results.push_back(result);
    }
    
    vector<bool> construct_candidates(vector<bool>&a, int i, int n, int k) {
        return vector<bool>{true, false};
    }
};
